metrics-exporter-prometheus/src/exporter/http_listener.rs

```diff
@@ -0,0 +1,131 @@
+use std::{
+    net::{IpAddr, SocketAddr},
+    sync::Arc,
+};
+
+use http_body_util::Full;
+use hyper::{
+    body::{self, Bytes, Incoming},
+    server::conn::http1::Builder as HyperHttpBuilder,
+    service::service_fn,
+    Request, Response, StatusCode,
+};
+use hyper_util::rt::TokioIo;
+use ipnet::IpNet;
+use tokio::net::{TcpListener, TcpStream};
+use tracing::warn;
+
+use crate::{common::BuildError, ExporterFuture, PrometheusHandle};
+
+struct HttpListeningExporter {
+    inner: Arc<Inner>,
+}
+#[derive(Clone)]
+struct Inner {
+    handle: PrometheusHandle,
+    allowed_addresses: Option<Vec<IpNet>>,
+}
+
+impl HttpListeningExporter {
+    async fn serve(&self, listener: std::net::TcpListener) -> Result<(), hyper::Error> {
+        // TODO - address panic possibility
+        let listener = TcpListener::from_std(listener).unwrap();
+
+        loop {
+            let stream = match listener.accept().await {
+                Ok((stream, _)) => stream,
+                Err(e) => {
+                    warn!("Error accepting connection. Ignoring request. Error: {:?}", e);
+                    continue;
+                }
+            };
+
+            let remote_addr = match stream.peer_addr() {
+                Ok(remote_address) => remote_address.ip(),
+                Err(e) => {
```

Comment:
The logic here makes sense, but I think what we might want to do is hoist up the logic from `handle_http_request`, for checking if the address is allowed, and then pass down the result of that into `process_stream`, and eventually, `handle_http_request`.

If we basically do something like...

```rust
let is_allowed = self.inner.allowed_addresses
    .map(|addrs| {
        stream.peer_addr().map(|peer_addr| {
            addrs.iter().any(|addr| addr.contains(&peer_addr))
        })
        .unwrap_or_else(|| {
            warn!("...");
            false
        })
    })
    .unwrap_or(true);
```

Then we always get `true` if no allowlist is configured, `false` if an allowlist _is_ configured but we can't detect the peer address, and whatever the result of the `any` call is if both are present.

Reply:
Sure  I was attempting to move the `is_allowed` check entirely into the `handle_http_request` method to offload from blocking the `serve` loop.  I also was trying to separate the three responsibilities:

 - `serve` = listen and loop
 - `process_stream` = spawn task
 - `handle_http_request` = authentication and processing
 
This, of course, necessitated the wrapping `allowed_addresses` withing the `Arc<Inner>`.  But this allowed me to only have to pass one server state var into `handle_http_request` along with the request state.  And this also adds a double `Arc` dereference to get to the recorder.

So a bunch of tradeoffs.  But your call.

Reply:
That makes sense, but I would say that, logically, checking if the remote address is  allowed should happen at the connection level, not per-request... even though, really, it's only going to be one request per connection in most cases.

I think the tradeoffs are worth it to get the level of specificity closer to optimal.